tests/e2e/websocket/websocket.cpp

```diff
@@ -0,0 +1,214 @@
+
+  WebsocketClient(Credentials creds) : session_{std::make_shared<Session>(ioc_, received_messages_, creds)} {}
+
+  void Connect(const std::string host, const std::string port) {
+    session_->Run(host, port);
+    bg_thread_ = std::thread([this]() { ioc_.run(); });
+    bg_thread_.detach();
+  }
+
+  void Close() { ioc_.stop(); }
+
+  std::vector<std::string> GetReceivedMessages() { return received_messages_; }
+
+ private:
+  std::vector<std::string> received_messages_{};
+  net::io_context ioc_{};
+  std::thread bg_thread_;
+  std::shared_ptr<Session> session_;
+};
+
+void TestWebsocketWithoutAnyUsers(auto &mg_client) {
+  spdlog::info("Starting websocket connection without any users.");
+  auto websocket_client = WebsocketClient();
+  websocket_client.Connect("127.0.0.1", "7444");
+
+  RunQueries(mg_client);
+  std::this_thread::sleep_for(std::chrono::seconds(1));
+
+  websocket_client.Close();
+  const auto received_messages = websocket_client.GetReceivedMessages();
+  spdlog::info("Received {} messages.", received_messages.size());
+  MG_ASSERT(!received_messages.empty(), "There are no received messages!");
+  for (const auto &log_message : received_messages) {
+    AssertLogMessage(log_message);
+  }
+  spdlog::info("Finishing websocket connection without any users.");
+}
+
+void TestWebsocketWithAuthentication(auto &mg_client) {
+  spdlog::info("Starting websocket connection with users.");
+  AddUser(mg_client);
+  std::this_thread::sleep_for(std::chrono::seconds(1));
+  auto websocket_client = WebsocketClient({"test", "testing"});
+  websocket_client.Connect("127.0.0.1", "7444");
+
+  RunQueries(mg_client);
+  std::this_thread::sleep_for(std::chrono::seconds(1));
+
+  websocket_client.Close();
+  const auto received_messages = websocket_client.GetReceivedMessages();
+  spdlog::info("Received {} messages.", received_messages.size());
+
+  MG_ASSERT(!received_messages.empty(), "There are no received messages!");
+  for (const auto &log_message : received_messages) {
+    AssertLogMessage(log_message);
+  }
+  spdlog::info("Finishing websocket connection with users.");
+}
+
+void TestWebsocketWithoutBeingAuthorized(auto &mg_client) {
+  spdlog::info("Starting websocket connection with users but without being authenticated.");
+  std::this_thread::sleep_for(std::chrono::seconds(1));
+  auto websocket_client = WebsocketClient();
+  websocket_client.Connect("127.0.0.1", "7444");
+
+  RunQueries(mg_client);
+  std::this_thread::sleep_for(std::chrono::seconds(1));
+
+  websocket_client.Close();
+  const auto received_messages = websocket_client.GetReceivedMessages();
+  spdlog::info("Received {} messages.", received_messages.size());
+
+  MG_ASSERT(received_messages.size() < 2, "There is no more than one received message!");
+  if (!received_messages.empty()) {
+    auto json_message = nlohmann::json::parse(received_messages[0]);
+    AssertAuthMessage(json_message, false);
+  }
```

Comment:
Why not expecting exactly one message? If the authentication fails, we expect the server to send a failed message, don't we?